test/Microsoft.Azure.ServiceBus.UnitTests/RetryTests.cs

```diff
@@ -0,0 +1,192 @@
+            ServerBusyException exception = new ServerBusyException(string.Empty);
+            TimeSpan retryInterval;
+
+            // First ServerBusy exception
+            Assert.False(policy1.IsServerBusy, "policy1.IsServerBusy should start with false");
+            Assert.True(policy1.ShouldRetry(duration, retryCount, exception, out retryInterval), "We should retry, but it returned false");
+            Assert.True(policy1.IsServerBusy, "policy1.IsServerBusy should be true");
+
+            System.Threading.Thread.Sleep(3000);
+
+            // Setting it a second time should not prolong the call.
+            Assert.True(policy1.IsServerBusy, "policy1.IsServerBusy should be true");
+            Assert.True(policy1.ShouldRetry(duration, retryCount, exception, out retryInterval), "We should retry, but it return false");
+            Assert.True(policy1.IsServerBusy, "policy1.IsServerBusy should be true");
+
+            System.Threading.Thread.Sleep(8000); // 3 + 8 = 11s
+            Assert.False(policy1.IsServerBusy, "policy1.IsServerBusy should stay false after 11s");
+
+            // Setting ServerBusy for second time.
+            Assert.True(policy1.ShouldRetry(duration, retryCount, exception, out retryInterval), "We should retry, but it return false");
+            Assert.True(policy1.IsServerBusy, "policy1.IsServerBusy is not true");
+        }
+
+        [Fact]
+        async void RunOperationShouldReturnImmediatelyIfRetryIntervalIsGreaterThanOperationTimeout()
+        {
+            var policy = RetryPolicy.Default;
+            Stopwatch watch = Stopwatch.StartNew();
+            try
+            {
+                await policy.RunOperation(() => { throw new ServiceBusException(true, string.Empty); }, TimeSpan.FromSeconds(3));
+            }
+            catch (Exception)
+            {
+                // Expected
+            }
+            Assert.True(watch.Elapsed.TotalSeconds < 3);
+        }
+
+        [Fact]
+        async void RunOperationShouldWaitFor10SecondsForOperationIfServerBusy()
+        {
+            var policy = RetryPolicy.Default;
+            policy.SetServerBusy(Resources.DefaultServerBusyException);
+            Stopwatch watch = Stopwatch.StartNew();
+            try
+            {
+                await policy.RunOperation(() => Task.CompletedTask, TimeSpan.FromMinutes(3));
+            }
+            catch (Exception)
+            {
+                // Expected
+            }
+            Assert.True(watch.Elapsed.TotalSeconds > 9);
+            Assert.False(policy.IsServerBusy);
+        }
+
+        [Fact]
+        async void RunOperationShouldWaitForAllOperationsToSucceed()
+        {
+            var policy = RetryPolicy.Default;
+            Stopwatch watch = Stopwatch.StartNew();
+            try
+            {
+                await policy.RunOperation(
+                    async () =>
+                    {
+                        for (int i = 0; i < 5; i++)
```

Comment:
@nemakam the link you're posted talks about the danger of passing an asyncy lambda into an `Action` will become a void call. The solution outlined within the post clearly states that the proper way is to use a `Func<Task>` instead of `Action`. And that is already done in your code and creating a loop doesn't change anything. Let me trace it here:

You have an async lambda passed into [`RetryPolicy.RunOperation()`](https://github.com/Azure/azure-service-bus-dotnet/pull/110/files/e93fef47f9d7fd1bf5ccff9d55d20d392dc09347#diff-c4777b8b45d84d6fbac7b96430200333R36) which _already_ takes in a `Func<Task>`. Internally within the method, you are [awaiting for the `operation`](https://github.com/Azure/azure-service-bus-dotnet/pull/110/files/e93fef47f9d7fd1bf5ccff9d55d20d392dc09347#diff-c4777b8b45d84d6fbac7b96430200333R59). 

So having multiple delays is not going to change anything and you could just have a single delay. If you would like to validate that several tasks got completed, you should have task handlers validated that after ` await policy.RunOperation()` is called, all tasks are in `Completed` state 😺